File: apps/mobile/src/__tests__/ingest-queue.test.ts
import { strict as assert } from "node:assert";
import { test } from "node:test";
import { randomUUID } from "node:crypto";
import { IngestQueue } from "../ingest/IngestQueue.js";
import type {
  BiosignalBatchEvent,
  OrganizationId,
  PatientId,
} from "@azri/contracts";

function fakeEvent(): BiosignalBatchEvent {
  const at = new Date().toISOString();
  return {
    type: "biosignal.batch",
    eventId: randomUUID(),
    schemaVersion: "0.2.0",
    occurredAt: at,
    organizationId: randomUUID() as OrganizationId,
    patientId: randomUUID() as PatientId,
    source: { deviceKind: "ios-phone", appVersion: "0.4.0" },
    phiClassification: "health-sensitive",
    kind: "heart_rate",
    unit: "bpm",
    samples: [{ at, value: 78 }],
  } as BiosignalBatchEvent;
}

test("flush dequeues only when post succeeds", async () => {
  let called = 0;
  const queue = new IngestQueue({
    poster: async () => {
      called++;
      return true;
    },
  });
  queue.enqueue([fakeEvent(), fakeEvent(), fakeEvent()]);
  const delivered = await queue.flush();
  assert.equal(delivered, 3);
  assert.equal(queue.depth(), 0);
  assert.equal(called, 1);
});

test("flush re-queues on failure", async () => {
  const queue = new IngestQueue({
    poster: async () => false,
  });
  queue.enqueue([fakeEvent(), fakeEvent()]);
  const delivered = await queue.flush();
  assert.equal(delivered, 0);
  assert.equal(queue.depth(), 2);
});

test("flush handles thrown errors as failure", async () => {
  const queue = new IngestQueue({
    poster: async () => {
      throw new Error("net");
    },
  });
  queue.enqueue([fakeEvent()]);
  const delivered = await queue.flush();
  assert.equal(delivered, 0);
  assert.equal(queue.depth(), 1);
});

File: apps/wear-os/app/src/test/java/ai/azri/wear/IngestQueueTest.kt
package ai.azri.wear

import ai.azri.wear.ingest.IngestQueue
import org.junit.Assert.assertEquals
import org.junit.Test
import java.time.Instant
import java.util.UUID

class IngestQueueTest {
    private fun event() = BiosignalBatchEvent(
        eventId = UUID.randomUUID().toString(),
        occurredAt = Instant.now().toString(),
        organizationId = UUID.randomUUID().toString(),
        patientId = UUID.randomUUID().toString(),
        source = EventSource(deviceKind = DeviceKind.WEAR_OS, appVersion = "0.4.0"),
        kind = BiosignalKind.HEART_RATE,
        unit = BiosignalUnit.BPM,
        samples = listOf(BiosignalSample(at = Instant.now().toString(), value = 78.0)),
        samplingHz = 1.0,
    )

    @Test fun flushSucceedsWhenPosterReturnsTrue() {
        val q = IngestQueue(maxBatchSize = 10)
        q.enqueue(listOf(event(), event(), event()))
        assertEquals(3, q.flush { true })
        assertEquals(0, q.depth())
    }

    @Test fun flushReQueuesOnFailure() {
        val q = IngestQueue(maxBatchSize = 10)
        q.enqueue(listOf(event(), event()))
        assertEquals(0, q.flush { false })
        assertEquals(2, q.depth())
    }

    @Test fun flushReQueuesOnException() {
        val q = IngestQueue(maxBatchSize = 10)
        q.enqueue(listOf(event()))
        assertEquals(0, q.flush { throw RuntimeException("network down") })
        assertEquals(1, q.depth())
    }
}
